x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java

```diff
@@ -0,0 +1,488 @@
+        new SimpleBatchedExecutor<>() {
+            @Override
+            public Tuple<ClusterState, Map<String, String>> executeTask(MarkRolesAsSyncedTask task, ClusterState clusterState) {
+                return task.execute(clusterState);
+            }
+
+            @Override
+            public void taskSucceeded(MarkRolesAsSyncedTask task, Map<String, String> value) {
+                task.success(value);
+            }
+        };
+
+    private final MasterServiceTaskQueue<MarkRolesAsSyncedTask> markRolesAsSyncedTaskQueue;
+
+    private final ClusterService clusterService;
+    private final FeatureService featureService;
+    private final QueryableBuiltInRoles.Provider rolesProvider;
+    private final SecurityIndexManager securityIndex;
+    private final QueryableBuiltInRolesStore queryableBuiltInRolesStore;
+    private final Executor executor;
+    private final AtomicBoolean synchronizationInProgress = new AtomicBoolean(false);
+
+    private volatile boolean securityIndexDeleted = false;
+
+    public QueryableBuiltInRolesSynchronizer(
+        ClusterService clusterService,
+        FeatureService featureService,
+        QueryableBuiltInRolesProviderFactory rolesProviderFactory,
+        QueryableBuiltInRolesStore queryableBuiltInRolesStore,
+        ReservedRolesStore reservedRolesStore,
+        FileRolesStore fileRolesStore,
+        SecurityIndexManager securityIndex,
+        ThreadPool threadPool
+    ) {
+        this.clusterService = clusterService;
+        this.featureService = featureService;
+        this.rolesProvider = rolesProviderFactory.createProvider(reservedRolesStore, fileRolesStore);
+        this.queryableBuiltInRolesStore = queryableBuiltInRolesStore;
+        this.securityIndex = securityIndex;
+        this.executor = threadPool.generic();
+        this.markRolesAsSyncedTaskQueue = clusterService.createTaskQueue(
+            "mark-built-in-roles-as-synced-task-queue",
+            Priority.LOW,
+            MARK_ROLES_AS_SYNCED_TASK_EXECUTOR
+        );
+        this.rolesProvider.addListener(this::builtInRolesChanged);
+        this.clusterService.addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void beforeStop() {
+                clusterService.removeListener(QueryableBuiltInRolesSynchronizer.this);
+            }
+
+            @Override
+            public void beforeStart() {
+                clusterService.addListener(QueryableBuiltInRolesSynchronizer.this);
+            }
+        });
+    }
+
+    private void builtInRolesChanged(QueryableBuiltInRoles roles) {
+        logger.debug("Built-in roles changed, attempting to sync to .security index");
+        final ClusterState state = clusterService.state();
+        if (shouldSyncBuiltInRoles(state)) {
+            syncBuiltInRoles(roles, state);
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
```

Comment:
>  Do you know if it'd be a huge pain to force write to index metadata to emulate roles changing in an IT? If it's not too difficult, we could then use that to e.g., delete a role or update one. That way we can test:

One way (that I can think of) to test this properly, is introducing a test plugin and registering `xpack.security.reserved_roles.include` setting. The changes to the setting will require a cluster restart, hence it will test other code paths (like .security index recovery).

Reply:
Pushed https://github.com/elastic/elasticsearch/pull/117581/commits/d400669b5a06f1bf3d7c2a99ce1ae7d4d8df9f61, which allows to test including and excluding roles (creating and deleting), but it does not help with testing a definition change (update). I think this is acceptable as we have this covered in the servereless repo.

Reply:
Also pushed https://github.com/elastic/elasticsearch/pull/117581/commits/cde5382551cb410eb2b7f1e1cafaef9ad7b64d41 which adds tests for 1., 2. and 3.